pkg/agent/cniserver/ipam/antrea_ipam.go

```diff
@@ -194,51 +211,68 @@ func (d *AntreaIPAM) Check(args *invoke.Args, k8sArgs *types.K8sArgs, networkCon
 	ipamConf := networkConfig.IPAM
 	numPools := len(ipamConf.IPPools)
-	if numPools == 0 {
-		return nil, fmt.Errorf("Antrea IPPool must be specified")
-	}
 
-	if err := d.waitForControllerReady(); err != nil {
-		// Return error to let the invoker retry.
-		return nil, err
+	if err := parseStaticAddresses(ipamConf); err != nil {
+		return nil, fmt.Errorf("failed to parse static addresses in the IPAM config: %v", err)
+	}
+	if numPools == 0 && len(ipamConf.Addresses) == 0 {
+		return nil, fmt.Errorf("at least one Antrea IPPool or static address must be specified")
 	}
-
-	owner := getAllocationOwner(args, k8sArgs, nil, true)
-	var allocatorsToRelease []*poolallocator.IPPoolAllocator
-	defer func() {
-		for _, allocator := range allocatorsToRelease {
-			// Try to release the allocated IPs after an error.
-			allocator.ReleaseContainer(owner.Pod.ContainerID, owner.Pod.IFName)
-		}
-	}()
 
 	result := &current.Result{}
-	for _, p := range ipamConf.IPPools {
-		allocator, err := d.controller.getPoolAllocatorByName(p)
-		if err != nil {
+	if numPools > 0 {
+		if err := d.waitForControllerReady(); err != nil {
+			// Return error to let the invoker retry.
 			return nil, err
 		}
 
-		var ip net.IP
-		var subnetInfo *crdv1a2.SubnetInfo
-		ip, subnetInfo, err = allocator.AllocateNext(crdv1a2.IPAddressPhaseAllocated, owner)
-		if err != nil {
-			return nil, err
-		}
-		if numPools > 1 {
-			allocatorsToRelease = append(allocatorsToRelease, allocator)
+		owner := getAllocationOwner(args, k8sArgs, nil, true)
+		var allocatorsToRelease []*poolallocator.IPPoolAllocator
+		defer func() {
+			for _, allocator := range allocatorsToRelease {
+				// Try to release the allocated IPs after an error.
+				allocator.ReleaseContainer(owner.Pod.ContainerID, owner.Pod.IFName)
+			}
+		}()
+
+		for _, p := range ipamConf.IPPools {
+			allocator, err := d.controller.getPoolAllocatorByName(p)
+			if err != nil {
+				return nil, err
+			}
+
+			var ip net.IP
+			var subnetInfo *crdv1a2.SubnetInfo
+			ip, subnetInfo, err = allocator.AllocateNext(crdv1a2.IPAddressPhaseAllocated, owner)
+			if err != nil {
+				return nil, err
+			}
+			if numPools > 1 {
+				allocatorsToRelease = append(allocatorsToRelease, allocator)
+			}
+
+			gwIP := net.ParseIP(subnetInfo.Gateway)
+			ipConfig, _ := generateIPConfig(ip, int(subnetInfo.PrefixLength), gwIP)
+			// CNI spec 0.2.0 and below support only one v4 and one v6 address. But we
+			// assume the CNI version >= 0.3.0, and so do not check the number of
+			// addresses.
+			result.IPs = append(result.IPs, ipConfig)
 		}
+		// No failed allocation, so do not release allocated IPs.
+		allocatorsToRelease = nil
+	}
 
-		gwIP := net.ParseIP(subnetInfo.Gateway)
-		ipConfig, _ := generateIPConfig(ip, int(subnetInfo.PrefixLength), gwIP)
-		result.IPs = append(result.IPs, ipConfig)
+	// Add static addresses.
+	for _, a := range ipamConf.Addresses {
```

Comment:
This method will not require an IPPool and antrea won't check this IP is allocated by others. Is it correct?

Reply:
What behavior we will get if user set both IPPool and IPAddress (in this IPPool or not in this IPPool)?

Reply:
Yes. We do not check and users need to make sure no IP conflicts. As described in #3634 , this is just what static IPAM plugin supports (https://www.cni.dev/plugins/current/ipam/static/) and it is also supported by Whereabouts. I added it for compatibility with them.

If both Pools and static addresses are specified, we will allocate IPs from Pools and append static addresses.